core/valor_core/classification.py

```diff
@@ -0,0 +1,1321 @@
+
+    if not (
+        enums.MetricType.PrecisionRecallCurve in metrics_to_return
+        or enums.MetricType.DetailedPrecisionRecallCurve in metrics_to_return
+    ):
+        return []
+
+    joint_df = (
+        pd.merge(
+            groundtruth_df,
+            prediction_df,
+            on=["datum_id", "datum_uid", "label_key"],
+            how="inner",
+            suffixes=("_gt", "_pd"),
+        ).assign(
+            is_label_match=lambda chain_df: (
+                (chain_df["label_value_pd"] == chain_df["label_value_gt"])
+            )
+        )
+        # only keep the columns we need
+        .loc[
+            :,
+            [
+                "datum_uid",
+                "datum_id",
+                "label_key",
+                "label_value_gt",
+                "id_gt",
+                "label_value_pd",
+                "score",
+                "id_pd",
+                "is_label_match",
+            ],
+        ]
+    )
+
+    # free up memory
+    del groundtruth_df
+    del prediction_df
+    gc.collect()
+
+    # add confidence_threshold to the dataframe and sort
+    pr_calc_df = pd.concat(
+        [
+            joint_df.assign(confidence_threshold=threshold)
+            for threshold in [x / 100 for x in range(5, 100, 5)]
+        ],
+        ignore_index=True,
+    ).sort_values(
+        by=[
+            "label_key",
+            "label_value_pd",
+            "confidence_threshold",
+            "score",
+        ],
+        ascending=False,
+    )
+
+    # create flags where the predictions meet criteria
+    pr_calc_df["true_positive_flag"] = (
+        pr_calc_df["score"] >= pr_calc_df["confidence_threshold"]
+    ) & pr_calc_df["is_label_match"]
+
+    # for all the false positives, we consider them to be a misclassification if they share a key but not a value with a gt
+    pr_calc_df["misclassification_false_positive_flag"] = (
+        pr_calc_df["score"] >= pr_calc_df["confidence_threshold"]
+    ) & ~pr_calc_df["is_label_match"]
+
+    # next, we flag false negatives by declaring any groundtruth that isn't associated with a true positive to be a false negative
+    groundtruths_associated_with_true_positives = (
+        pr_calc_df[pr_calc_df["true_positive_flag"]]
+        .groupby(["confidence_threshold"], as_index=False)["id_gt"]
+        .unique()
+    )
+
+    if not groundtruths_associated_with_true_positives.empty:
+        confidence_interval_to_true_positive_groundtruth_ids_dict = (
+            groundtruths_associated_with_true_positives.set_index(
+                "confidence_threshold"
+            )["id_gt"]
+            .apply(set)
+            .to_dict()
+        )
+
+        pr_calc_df["false_negative_flag"] = pr_calc_df.apply(
+            lambda row: row["id_gt"]
+            not in confidence_interval_to_true_positive_groundtruth_ids_dict.get(
+                row["confidence_threshold"], set()
+            ),
+            axis=1,
+        )
```

Comment:
```suggestion
        false_negative_mask = pd.Series(False, index=pr_calc_df.index)
        for threshold, elements in confidence_interval_to_true_positive_groundtruth_ids_dict.items():
            threshold_mask = pr_calc_df["confidence_threshold"] == threshold
            membership_mask = ~pr_calc_df["id_gt"].isin(elements)
            false_negative_mask |= threshold_mask & membership_mask
        pr_calc_df["false_negative_flag"] = false_negative_mask
```

After the 9e606cd changes, this line and the corresponding one below account for a large majority of the running time of this function.   I found a large speedup (and decreased memory usage) by using array functions as suggested.